lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,543 @@
+
+use super::{Collection, Kind};
+use lookup::{Lookup, Segment};
+
+impl Kind {
+    /// Find the [`Kind`] at the given path.
+    ///
+    /// If the path points to root, then `self` is returned, otherwise `None` is returned if `Kind`
+    /// isn't an object or array, or if the path points to a non-existing field/index in the
+    /// object/array.
+    ///
+    /// Negative indexing always returns `None`, as the type system doesn't know the exact size of
+    /// an array, and thus cannot count backward.
+    #[must_use]
+    pub fn find_at_path(&self, path: &Lookup<'_>) -> Option<&Kind> {
+        if path.is_root() {
+            return Some(self);
+        }
+
+        let mut kind = self;
+        for segment in path.iter() {
+            kind = match segment {
+                // Try finding the field in the existing object.
+                Segment::Field(field) => kind
+                    .object
+                    .as_ref()
+                    .and_then(|collection| collection.known().get(&(field.into())))?,
+
+                // Try finding one of the fields in the existing object.
+                Segment::Coalesce(fields) => kind.object.as_ref().and_then(|collection| {
+                    let field = fields
+                        .iter()
+                        .find(|field| collection.known().contains_key(&((*field).into())))?;
+
+                    collection.known().get(&(field.into()))
+                })?,
+
+                // Try finding the index in the existing array.
+                Segment::Index(index) => usize::try_from(*index).ok().and_then(|index| {
+                    kind.array
+                        .as_ref()
+                        .and_then(|collection| collection.known().get(&(index.into())))
+                })?,
+            };
+        }
+
+        Some(kind)
+    }
+
+    /// Nest the given [`Kind`] into a provided path.
+    ///
+    /// For example, given an `integer` kind and a path `.foo`, a new `Kind` is returned that is
+    /// known to be an object, of which the `foo` field is known to be an `integer`.
+    #[must_use]
+    pub fn nest_at_path(mut self, path: &Lookup<'_>) -> Self {
+        fn object_from_field(field: &lookup::Field<'_>, kind: Kind) -> Kind {
+            let map = BTreeMap::from([(field.into(), kind)]);
+            Kind::object(map)
+        }
+
+        for segment in path.iter().rev() {
+            match segment {
+                Segment::Field(field) => {
+                    self = object_from_field(field, self);
+                }
+                Segment::Coalesce(fields) => {
+                    // We pick the last field in the list of coalesced fields, there is no
+                    // "correct" way to handle this case, other than not supporting it, or making
+                    // this method call fallible.
+                    let field = fields.last().expect("at least one");
+                    self = object_from_field(field, self);
```

Comment:
Couldn't we add the kind to all fields indicated in the coalesce?

Reply:
We could, but I wouldn't actually call that "correct" either.

I consider coalesced paths to be "start from the first, and progress until one matches", and when inserting, I consider it to be "take the first, no matter what" (which I realize is different from this, so I should update this).

Reply:
I decided to reject coalesced nesting for now. We can revisit this later.